File: PtolC/sensor.c
/*
 * PtolC/sensor.c — Physical sensor bridge: reads live_state.json.
 *
 * Maps 8 sensor channels (e₀..e₇, lower 𝕆 operator labels) from
 * ~/.ptolemy/live_state.json into a float vector for field ingestion.
 *
 * JSON format accepted:
 *   {"sensor": [v0, v1, v2, v3, v4, v5, v6, v7]}     (array form)
 *   {"sensor": {"identity": v, "negate": v, ...}}     (named form)
 *   {"sensor": {"0": v, "1": v, ...}}                 (index-string form)
 *
 * Channel → lower-𝕆 operator:
 *   0 identity  1 negate  2 bind  3 name
 *   4 apply     5 abstract  6 branch  7 iterate
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "sensor.h"
#include "log.h"

/* Default state file path (expanded at runtime) */
#define SENSOR_PATH_MAX 4096

static const char *CHANNEL_NAMES[8] = {
    "identity", "negate", "bind", "name",
    "apply", "abstract", "branch", "iterate"
};

/* ── Path resolution ──────────────────────────────────────────────────────── */

static void sensor_default_path(char *buf, size_t n)
{
    const char *home = getenv("HOME");
    if (!home) home = "/root";
    snprintf(buf, n, "%s/.ptolemy/live_state.json", home);
}

/* ── Minimal JSON float extractor ────────────────────────────────────────── */

/*
 * Find "key": value in a JSON string.
 * Returns 1 on success, 0 if not found.
 * For string keys, returns the numeric value following ":".
 */
static int json_get_float(const char *src, const char *key, double *out)
{
    char pat[64];
    snprintf(pat, sizeof(pat), "\"%s\":", key);
    const char *p = strstr(src, pat);
    if (!p) return 0;
    p += strlen(pat);
    while (*p == ' ' || *p == '\t') p++;
    char *end;
    double v = strtod(p, &end);
    if (end == p) return 0;
    *out = v;
    return 1;
}

/* ── Public API ───────────────────────────────────────────────────────────── */

int sensor_read(float *channels_out, const char *path)
{
    char default_path[SENSOR_PATH_MAX];
    if (!path || !path[0]) {
        sensor_default_path(default_path, sizeof(default_path));
        path = default_path;
    }

    for (int i = 0; i < 8; i++) channels_out[i] = 0.0f;

    FILE *f = fopen(path, "r");
    if (!f) {
        plog(PLOG_WARN, "sensor_read: cannot open %s", path);
        return 0;
    }

    /* Read entire file */
    fseek(f, 0, SEEK_END);
    long sz = ftell(f);
    fseek(f, 0, SEEK_SET);
    if (sz <= 0 || sz > 65536) { fclose(f); return 0; }

    char *buf = malloc((size_t)sz + 1);
    if (!buf) { fclose(f); return 0; }
    fread(buf, 1, (size_t)sz, f);
    fclose(f);
    buf[sz] = '\0';

    /* Locate "sensor" key */
    const char *sens = strstr(buf, "\"sensor\"");
    if (!sens) { free(buf); return 0; }
    sens += 8;   /* skip "sensor" */
    while (*sens == ':' || *sens == ' ') sens++;

    if (*sens == '[') {
        /* Array form: parse up to 8 floats */
        const char *p = sens + 1;
        for (int i = 0; i < 8; i++) {
            while (*p == ' ' || *p == ',') p++;
            if (*p == ']') break;
            char *end;
            double v = strtod(p, &end);
            if (end == p) break;
            channels_out[i] = (float)v;
            p = end;
        }
    } else if (*sens == '{') {
        /* Object form: named or index-string keys */
        for (int i = 0; i < 8; i++) {
            double v = 0.0;
            char idx[4];
            snprintf(idx, sizeof(idx), "%d", i);
            if (json_get_float(sens, CHANNEL_NAMES[i], &v) ||
                json_get_float(sens, idx, &v)) {
                channels_out[i] = (float)v;
            }
        }
    }

    free(buf);
    return 1;
}

int sensor_write(const float *channels, const char *path)
{
    char default_path[SENSOR_PATH_MAX];
    if (!path || !path[0]) {
        sensor_default_path(default_path, sizeof(default_path));
        path = default_path;
    }

    FILE *f = fopen(path, "w");
    if (!f) return 0;
    fprintf(f, "{\n  \"sensor\": {\n");
    for (int i = 0; i < 8; i++) {
        fprintf(f, "    \"%s\": %.6f%s\n",
                CHANNEL_NAMES[i], (double)channels[i],
                i < 7 ? "," : "");
    }
    fprintf(f, "  }\n}\n");
    fclose(f);
    return 1;
}

void sensor_print(const float *channels, FILE *out)
{
    /* L2 norm for callosum coupling proxy */
    double norm = 0.0;
    for (int i = 0; i < 8; i++)
        norm += (double)channels[i] * (double)channels[i];
    norm = sqrt(norm);

    fprintf(out, "sensor_channels:\n");
    for (int i = 0; i < 8; i++) {
        fprintf(out, "  e%-2d %-10s %.6f\n",
                i, CHANNEL_NAMES[i], (double)channels[i]);
    }
    fprintf(out, "  L2_norm %.6f\n", norm);
}

File: skills/sensor.py
"""
skills/sensor.py — Physical sensor bridge: e₀..e₇ (first 𝕆 / lower half).

:description:
    SensorReader maps 8 live sensor channels onto the first octonion copy
    of the sedenion field (e₀..e₇) via MindEye.see().

    The 8 channels mirror the lower-𝕆 operator labels::

        e₀  identity   — raw scalar / amplitude
        e₁  negate     — inverse / phase-flip signal
        e₂  bind       — coupling strength / coherence
        e₃  name       — source label hash (normalised)
        e₄  apply      — transform output / activation
        e₅  abstract   — spectral envelope / frequency centroid
        e₆  branch     — bifurcation index / modal split
        e₇  iterate    — temporal phase / cycle counter

    Data sources (all read from ``~/.ptolemy/live_state.json``):
      - Manual write by external process (e.g. microphone capture, GPIO)
      - MIT Dolphin acoustic channels (24D → 8D projection, pending)
      - Any 8-float vector keyed under ``"sensor"`` in the JSON

    The zero-divisor bridge means sensor data in e₀..e₇ reaches
    cognitive dims e₈..e₁₅ only through the 42 Cawagas callosum pairs.
    SensorReader feeds that lower half so the upper half can respond.

:classes:
    SensorReader

:constants:
    CHANNEL_NAMES, DEFAULT_STATE_PATH
"""

import json
import math
import os
import threading
import time
from typing import Any, Callable, Dict, List, Optional

DEFAULT_STATE_PATH = os.path.expanduser('~/.ptolemy/live_state.json')

# Lower-𝕆 operator labels (mirrors draw.py _OP)
CHANNEL_NAMES = {
    0: 'identity',
    1: 'negate',
    2: 'bind',
    3: 'name',
    4: 'apply',
    5: 'abstract',
    6: 'branch',
    7: 'iterate',
}


class SensorReader:
    """
    Physical sensor bridge for the lower octonion (e₀..e₇).

    Reads ``~/.ptolemy/live_state.json``, extracts 8 channel values,
    and forwards them into ``MindEye.see()`` so the cognitive half
    (e₈..e₁₅) can respond via zero-divisor callosum coupling.

    :param engine: Live ``Engine`` instance.
    :param state_path: Path to ``live_state.json``. Defaults to
        ``~/.ptolemy/live_state.json``.

    :Example:

    .. code-block:: python

        sr = engine.get_sensor_reader()
        result = sr.see()
        print(result['callosum'], result['channels'])

        # Start background poll loop
        sr.watch(interval=1.0)
        # ... later ...
        sr.stop()
    """

    def __init__(self, engine, state_path: str = DEFAULT_STATE_PATH):
        self._engine     = engine
        self._state_path = state_path
        self._lock       = threading.Lock()
        self._watching   = False
        self._thread: Optional[threading.Thread] = None
        self._last_read: Dict[str, Any] = {}
        self._callbacks: List[Callable[[Dict[str, Any]], None]] = []

    # ── I/O ──────────────────────────────────────────────────────────────────

    def read(self) -> Dict[str, float]:
        """
        Read 8 sensor channels from ``live_state.json``.

        Returns a dict keyed by channel name. Missing channels default to 0.0.
        If the file does not exist or cannot be parsed, returns all zeros.

        :returns: Dict mapping channel name → float value.
        :rtype: dict
        """
        channels: Dict[str, float] = {name: 0.0 for name in CHANNEL_NAMES.values()}
        if not os.path.exists(self._state_path):
            return channels
        try:
            with open(self._state_path, 'r') as f:
                data = json.load(f)
        except Exception:
            return channels

        sensor = data.get('sensor', {})
        if isinstance(sensor, list):
            # list form: [v0, v1, ..., v7]
            for i, v in enumerate(sensor[:8]):
                if i in CHANNEL_NAMES:
                    channels[CHANNEL_NAMES[i]] = float(v)
        elif isinstance(sensor, dict):
            # dict form: {'identity': 0.3, 'negate': 0.7, ...}
            # also accepts integer keys '0'..'7'
            for k, v in sensor.items():
                if k in CHANNEL_NAMES.values():
                    channels[k] = float(v)
                elif str(k).isdigit():
                    idx = int(k)
                    if idx in CHANNEL_NAMES:
                        channels[CHANNEL_NAMES[idx]] = float(v)

        with self._lock:
            self._last_read = dict(channels)
        return channels

    def write(self, channels: Dict[str, float]) -> bool:
        """
        Write sensor channels to ``live_state.json`` (creates if absent).

        Only writes the ``"sensor"`` key; preserves all other top-level keys.

        :param channels: Dict mapping channel name or int index → value.
        :returns: True on success, False on error.
        :rtype: bool
        """
        data: Dict[str, Any] = {}
        if os.path.exists(self._state_path):
            try:
                with open(self._state_path, 'r') as f:
                    data = json.load(f)
            except Exception:
                pass
        # Normalise to name-keyed dict
        norm: Dict[str, float] = {}
        for k, v in channels.items():
            if k in CHANNEL_NAMES.values():
                norm[k] = float(v)
            elif str(k).isdigit():
                idx = int(k)
                if idx in CHANNEL_NAMES:
                    norm[CHANNEL_NAMES[idx]] = float(v)
        data['sensor'] = norm
        try:
            os.makedirs(os.path.dirname(self._state_path), exist_ok=True)
            with open(self._state_path, 'w') as f:
                json.dump(data, f, indent=2)
            return True
        except Exception:
            return False

    # ── field ingestion ───────────────────────────────────────────────────────

    def see(self) -> Dict[str, Any]:
        """
        Read sensor state and encode into MindEye (e₈..e₁₅).

        The 8 channel values become the second-𝕆 vector. The callosum
        coupling strength (e₁₅) is computed by MindEye from the vector norm
        vs. OMEGA_ZS. High coupling = sensor state near the neutral buoyancy
        surface = maximum zero-divisor channel activation.

        :returns: Dict with keys ``channels``, ``callosum``, ``heard``.
        :rtype: dict
        """
        channels = self.read()
        vec      = [channels[CHANNEL_NAMES[i]] for i in range(8)]

        # Normalise vector to unit range for MindEye
        norm = math.sqrt(sum(v * v for v in vec))
        if norm > 0:
            vec = [v / norm for v in vec]

        me   = self._engine.get_mind_eye()
        seen = me.see(vec, label='sensor')
        heard = ''
        try:
            # Synthesise a description from the strongest channel
            top_dim = max(range(8), key=lambda i: abs(vec[i]))
            prompt  = CHANNEL_NAMES[top_dim]
            heard   = self._engine.hear(prompt)
        except Exception:
            pass
        return {'channels': channels, 'vec': vec,
                'callosum': seen.get('callosum', 0.0), 'heard': heard}

    # ── poll loop ─────────────────────────────────────────────────────────────

    def on_update(self, callback: Callable[[Dict[str, Any]], None]):
        """
        Register a callback invoked on each successful sensor read during watch().

        :param callback: Callable receiving the result dict from ``see()``.
        """
        self._callbacks.append(callback)

    def watch(self, interval: float = 1.0):
        """
        Start background poll loop. Calls ``see()`` every ``interval`` seconds.

        Non-blocking — spawns a daemon thread. Safe to call multiple times
        (second call is a no-op if already watching).

        :param interval: Poll interval in seconds.
        """
        if self._watching:
            return
        self._watching = True

        def _loop():
            while self._watching:
                try:
                    result = self.see()
                    for cb in list(self._callbacks):
                        try:
                            cb(result)
                        except Exception:
                            pass
                except Exception:
                    pass
                time.sleep(interval)

        self._thread = threading.Thread(target=_loop, daemon=True,
                                        name='SensorReader')
        self._thread.start()

    def stop(self):
        """
        Stop the background poll loop.

        :returns: ``{'watching': False}``.
        :rtype: dict
        """
        self._watching = False
        if self._thread:
            self._thread.join(timeout=3.0)
            self._thread = None
        return {'watching': False}

    def status(self) -> Dict[str, Any]:
        """
        Return current watcher status and last read channels.

        :returns: Dict with keys ``watching``, ``last_read``, ``path``.
        :rtype: dict
        """
        with self._lock:
            last = dict(self._last_read)
        return {'watching': self._watching, 'last_read': last,
                'path': self._state_path}
